### PR TITLE
Make QuasiMonotoneLimiter work on GPU and MPI

### LINKED ISSUE #2610: Restarts with MPI and QuasiMonotoneLimiter do not work
## Describe the bug
An `isbits` error is thrown when trying to restart an mpi simulation in ClimaAtmos that uses the QuasiMonotoneLimiter. This is not exercised in ClimaCore or ClimaAtmos ci.

## To Reproduce

I have not had time to make a minimal reproducer. [This](https://github.com/CliMA/ClimaAtmos.jl/commit/f1e626160292e2981e18ba9cb8b29626cf01bad4) is a commit on a ClimaAtmos branch that can reproduce this bug.

### PR DESCRIPTION
**Pull request by ph-kev**
closes https://github.com/CliMA/ClimaCore.jl/issues/2610 - This PR fixes the issue by adding support for CUDA + MPI for the `QuasiMonotoneLimiter`.

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -613,8 +613,10 @@ steps:
         command: "srun julia --color=yes --check-bounds=yes --project=.buildkite test/Limiters/distributed/dlimiter.jl"
         env:
           CLIMACOMMS_CONTEXT: "MPI"
+          CLIMACOMMS_DEVICE: "CUDA"
```
Comment (imreddyTeja) [inline .buildkite/pipeline.yml:616]: probably a good idea to keep an MPI + CPU test instead of replacing it

File: src/Limiters/quasimonotone.jl [hunk 2]
```diff
@@ -248,19 +248,21 @@ end
 Update the field `limiter.q_bounds_nbr` based on `limiter.q_bounds` in the ghost
 neighbors. This should be called after the ghost exchange has completed.
```
Comment (imreddyTeja) [inline src/Limiters/quasimonotone.jl:250]: I think the `min_tracer_mass = Inf` in `apply_limiter!` below might be a bug. Could you look into that if possible? It would also be good to update the PR description with a list of things fixed

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (ph-kev) [reply .buildkite/pipeline.yml:616]: Do we want to keep both MPI + GPU and MPI + CPU tests?

Comment (imreddyTeja) [reply .buildkite/pipeline.yml:616]: up to you. you probably know the most about this now

File: src/Limiters/quasimonotone.jl (hunk 2 above)
Comment (tapios) [reply src/Limiters/quasimonotone.jl:250]: Yes, it's a bug. It is passed into a max, that then always defaults to Inf. Would be good to fix here.

**Review by tapios — COMMENTED** [review]
A few other bugs in limiters that would be good to fix in this PR:

1) quasimonotone.jl (l. 379). The warn is unconditional and warns whenever the `warn` flag is present. Should be conditional on non-convergence: `if warn && !converged`

2) `apply_limiter!` on GPU returns `nothing` (ext/cuda/limiters.jl, l. 145); on CPU it returns `ρq`. Would be good to make consistent (I'm not sure why `ρq` is returned on CPU, maybe it can also be nothing?).

3) `QuasiMonotoneLimiter` should probably a subtype of `AbstractLimiter`, like the other limiters, i.e., `struct QuasiMonotoneLimiter{D, G, FT, CS} <: AbstractLimiter` in quasimonotone.jl, l. 98.

**Comment by tapios** [comment]
> A few other bugs in limiters that would be good to fix in this PR:
> 
> 1. quasimonotone.jl (l. 379). The warn is unconditional and warns whenever the `warn` flag is present. Should be conditional on non-convergence: `if warn && !converged`
> 2. `apply_limiter!` on GPU returns `nothing` (ext/cuda/limiters.jl, l. 145); on CPU it returns `ρq`. Would be good to make consistent (I'm not sure why `ρq` is returned on CPU, maybe it can also be nothing?).
> 3. `QuasiMonotoneLimiter` should probably a subtype of `AbstractLimiter`, like the other limiters, i.e., `struct QuasiMonotoneLimiter{D, G, FT, CS} <: AbstractLimiter` in quasimonotone.jl, l. 98.

Correction to previous version of the above comment: The quasi-monotone limiter must be fixed in a different way. 

a) Change quasimonotone.jl (l. 372) to `min_tracer_mass = min(min_tracer_mass, slab_min_tracer_mass)`
b) Change the logic in quasimonotone.jl (l. 502 ff)

```
converged_now = abs(Δtracer_mass) <= rtol * abs(tracer_mass)
rel_err = iszero(tracer_mass) ? zero(FT) : abs(Δtracer_mass) / abs(tracer_mass)
max_rel_err = max(max_rel_err, rel_err)
min_tracer_mass = min(min_tracer_mass, abs(tracer_mass))
converged_now && break
```

The problem is the division by the tracer mass in the current code. It NaNs for zero tracer, which is not uncommon, e.g., for specific humidity. 

The per-iteration update of `min_tracer_mass` inside the loop is also redundant since tracer_mass is constant for the slab, and can be hoisted above the loop.

@akshaysridhar can help here. Would be good add/revise tests so this works reliably (e.g., add a unit test that runs the limiter on a field with an all-zero element and asserts convergence and a finite `max_rel_err`--it would have caught this bug).
